Player firing

jmortiger opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)
Main camera not correctly initialized for PlayArea

## Type of change

Refactored color changing properties to a base class
Player implements ColoredObj base class
Bullet implements ColoredObj base class
Player shoots bullets

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update


## How were the changes tested

Describe how you tested the changes. So others can replicate your tests and confirm similar results on their machine.

Tested usage with Unity. Attempted moving, colorchanging, and firing. Bullets are the color of the ship, have velocity, and spawn near tip of ship.
One known error with meshRender init. was worked around. I don't see it being a problem but it's something to hopefully fix later on.

## Self-Review Checklist:

- [x] Code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors

jmortiger committed:
Switched player prefab model
Made temp bullet prefab
Fixed small PlayerArea camera bug

jmortiger committed:
Began adding firing

jmortiger committed:
progress

Commit:
Changed hierarchy to PlayerShip & Bullet inheriting from ColoredObj
Got firing working
Got Bullet colors working

jmortiger committed:
Self-Review work

jmortiger committed:
Merge branch 'master' into PlayerFiring

jmortiger commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Use Player Ship model & add firing](https://trello.com/c/p4b2T4Lm/17-use-player-ship-model-add-firing)

EthanNichols commented on Digital 1/Assets/Scripts/Bullet.cs:
```diff
@@ -0,0 +1,42 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Digital1
```
Get rid of namespaces. Project isn't big enough for them to be reasonably used.

jmortiger replied:
Gotcha, was trying their use for a bug fix and forgot to remove them.

EthanNichols commented on Digital 1/Assets/Scripts/ColoredObj.cs:
```diff
@@ -0,0 +1,91 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Digital1
```
Remove namespace.

EthanNichols commented on Digital 1/Assets/Scripts/Player/PlayerShip.cs:
```diff
@@ -1,120 +1,97 @@
 using UnityEngine;
 
-public class PlayerShip : MonoBehaviour
+namespace Digital1
```
Remove namespace

EthanNichols commented on Digital 1/Assets/Scripts/ColoredObj.cs:
```diff
@@ -0,0 +1,91 @@
+        /// Handle Color State switching, then update Material/Mesh.
+        /// </summary>
+        /// <param name="colorIndex"></param>
+        protected void ColorSwitch(int colorIndex)
```
This should use the ColorState enum as a parameter instead of an int.

jmortiger replied:
I left that from the original pulled code, but I agree, and think that was a change from the revised code I forgot to add.

EthanNichols commented on Digital 1/Assets/Scripts/ColoredObj.cs:
```diff
@@ -0,0 +1,91 @@
+
+        void UpdateMaterial()
+        {
+            currentMaterial = colorRef[(int)currentState];
```
ColorState enum values should correlate directly to a number. Rather than assuming that the color starts at 0.

EthanNichols replied:
Should also probably add a check to ensure that the index being grabbed exists.

jmortiger replied:
I changed the ColorState definition to manually define values, but I'm not sure if this is what you meant.
`public enum ColorState
    {
        Neutral = 0,
        Red     = 1,
        Blue    = 2,
        Yellow  = 3
    }`
I also added the check.
```
// currentMaterial = colorRef?[(int)currentState] ?? currentMaterial; // I think Unity has null operator problems, not sure. If not, this is better.
try
{
     currentMaterial = colorRef[(int)currentState];
}
catch (System.IndexOutOfRangeException)
{
     Debug.LogWarning("IndexOutOfRangeException in ColoredObj.UpdateMaterial(). Failed to change Material.");
}```

EthanNichols replied:
Ya that's what I meant. It's just safer to know the value has been set in the enum rather than assuming that it'll work.

EthanNichols reviewed (changes requested):
There are some weird formatting issues with your scripts.

Try to reduce the amount of code that you move around in a file. Playship.cs seems like it was completely rewritten, when in reality everything was just moved.

jmortiger committed:
Removed States (Is now redundant)
Removed Namespaces
Fixed Formatting Issues from Github's editor
Added explict enum defs and index error checks

EthanNichols reviewed (approved):
Besides some minor formatting issues looks good.
